### PR TITLE
[WIP] Added MatrixNormal

### PR DESCRIPTION
**Pull request by jordan-melendez**
This PR adds the functionality discussed [here](https://discourse.pymc.io/t/multiple-uncertain-function-observations-of-the-same-gaussian-process/400/15). This is marked as WIP because
* the `shape` kwarg is currently required, must specify the shape of the qxp matrix
* it does not support multiple draws from the distribution, i.e., `len(shape) > 2`
* there is no example in the docstring
* it has not been extensively tested
* there may be better/quicker ways to implement it
* etc.

I used the following as guides:
* [Wikipedia](https://en.wikipedia.org/wiki/Matrix_normal_distribution)
* [This article](http://ftp.stat.duke.edu/WorkingPapers/08-01.pdf)

Any help getting this up to your standards would be appreciated! Thanks.

### CODE REVIEW

File: pymc3/distributions/multivariate.py [hunk 1]
```diff
@@ -1091,3 +1091,183 @@ def logp(self, x):
...
+    --------
+    ?
+    """
+
+    def __init__(self, mu=0, rcov=None, rchol=None, rtau=None,
+                 lcov=None, lchol=None, ltau=None, *args, **kwargs):
+
+        self.setup_matrices(rcov, rchol, rtau, lcov, lchol, ltau)
+
+        shape = kwargs.pop('shape', None)
```
Comment (aseyboldt) [inline pymc3/distributions/multivariate.py]: I don't see a problem in making the requiring the shape argument. Just raise a `TypeError` if you don't get one. (I guess in py3 we could make that a required keyword-only argument...)

File: pymc3/distributions/multivariate.py [hunk 2]
```diff
@@ -1091,3 +1091,183 @@ def logp(self, x):
...
+        self.shape = shape
+
+        super(MatrixNormal, self).__init__(shape=shape, *args, **kwargs)
+
+        self.mu = tt.as_tensor_variable(mu)
+
+        self.mean = self.median = self.mode = self.mu
+
+        # self.solve_lower = tt.slinalg.Solve(A_structure="lower_triangular")
+        # self.solve_upper = tt.slinalg.Solve(A_structure="upper_triangular")
```
Comment (aseyboldt) [inline pymc3/distributions/multivariate.py]: Those were a work around from before the theano 0.9 release. I think we can safely delete those.

File: pymc3/distributions/multivariate.py [hunk 3]
```diff
@@ -1091,3 +1091,183 @@ def logp(self, x):
...
+        self.solve_lower = tt.slinalg.solve_lower_triangular
+        self.solve_upper = tt.slinalg.solve_upper_triangular
+
+    def setup_matrices(self, rcov, rchol, rtau, lcov, lchol, ltau):
+        # Step methods and advi do not catch LinAlgErrors at the
```
Comment (aseyboldt) [inline pymc3/distributions/multivariate.py]: NUTS should catch LinAlgErrors now. Maybe check how much faster things get if we set `nofail=False` and also get rid of the related nan checks later. advi stops if it runs in nan anyway...

File: pymc3/distributions/multivariate.py [hunk 4]
```diff
@@ -1091,3 +1091,183 @@ def logp(self, x):
...
+    rchol : array
+        Cholesky decomposition of pxp covariance matrix. Defines variance
+        within rows. Exactly one of rcov or rchol is needed.
+    lcov : array
+        Left (or column) qxq covariance matrix. Defines variance within
+        columns. Exactly one of lcov or lchol is needed.
+    lchol : array
+        Cholesky decomposition of qxq covariance matrix. Defines variance
+        within columns. Exactly one of lcov or lchol is needed.
+
```
Comment (aseyboldt) [inline pymc3/distributions/multivariate.py]: I'd list the shape explicitly here, I think it could use a bit of explanation with all the shapes floating around.

File: pymc3/distributions/multivariate.py [hunk 5]
```diff
@@ -1091,3 +1091,183 @@ def logp(self, x):
...
+        if size is None:
+            size = list(self.shape)
+
+        mu, rchol, lchol = draw_values(
+                                [self.mu, self.rchol_cov, self.lchol_cov],
+                                point=point
+                                )
+        standard_normal = np.random.standard_normal(size)
+
+        return mu + lchol @ standard_normal @ rchol.T
```
Comment (aseyboldt) [inline pymc3/distributions/multivariate.py]: This isn't py2 compatible

File: pymc3/distributions/multivariate.py [hunk 6]
```diff
@@ -1091,3 +1091,183 @@ def logp(self, x):
...
+
+        lchol_cov = self.lchol_cov
+        rchol_cov = self.rchol_cov
+
+        rdiag = tt.nlinalg.diag(rchol_cov)
+        ldiag = tt.nlinalg.diag(lchol_cov)
+        # Check if the covariance matrix is positive definite.
+        rok = tt.all(rdiag > 0)
+        lok = tt.all(ldiag > 0)
+        ok = rok and lok
```
Comment (aseyboldt) [inline pymc3/distributions/multivariate.py]: I don't think this `and` does the right thing. There is `tt.and_`.

**Review by aseyboldt — COMMENTED** [review]
This look good, but I think it could use a bit more explanation and an example in the docstring. Also some tests against MvNormal to see if it actually gives the right answers would be helpful.

File: pymc3/distributions/multivariate.py [hunk 7]
```diff
@@ -1091,3 +1091,219 @@ def logp(self, x):
...
+
+    def __init__(self, mu=0, rcov=None, rchol=None, rtau=None,
+                 lcov=None, lchol=None, ltau=None, *args, shape=None,
+                 **kwargs):
+
+        self._setup_matrices(rcov, rchol, rtau, lcov, lchol, ltau)
+
+        if shape is None:
+            raise TypeError('shape is a required argument')
+        assert len(shape) == 2, "shape must have length 2: qxp"
```
Comment (bwengals) [inline pymc3/distributions/multivariate.py]: I think it would be better to raise a `ValueError` here instead of an `AssertionError`

Comment (bwengals) [reply pymc3/distributions/multivariate.py]: Also, if there are multiple observations, should the shape be `q x p x n_obs`?

**Comment by bwengals** [comment]
Nice work!  For testing, I think you could do something[ like here](https://github.com/pymc-devs/pymc3/blob/master/pymc3/tests/test_distributions.py#L639), against [SciPy's matrix normal](https://docs.scipy.org/doc/scipy-0.18.1/reference/generated/scipy.stats.matrix_normal.html).

**Comment by jordan-melendez** [comment]
I did not see that SciPy had a matrix normal, so I can try to do that. Earlier I wrote up this test against MvNormal:

```python
def test_matrixnormal_vs_mvnormal():
    q = 5
    p = 2

    # Test logp(x) for both distributions
    x = tt.matrix('x')
    x.tag.test_value = np.zeros((q, p))
    x_test = np.ones((q, p))

    # Setup mean, standard covariance matrix and MvNormal logp
    mu = floatX(np.zeros((q, p)))
    cov_val = np.array([[1, 0.5], [0.5, 1]])
    assert cov_val.shape == (p, p)
    cov = tt.matrix('cov')
    cov.tag.test_value = np.eye(p)
    logp = pm.MvNormal.dist(mu=mu, cov=cov, shape=(q, p)).logp(x)
    f_logp = theano.function([cov, x], logp)
    mv_logp_value = f_logp(cov_val, x_test).sum()

    # Setup left covariance matrix (identity), and ensure MatrixNormal
    # logp reduces to MvNormal logp
    lcov_val = np.eye(q)
    lcov = tt.matrix('lcov')
    lcov.tag.test_value = np.eye(q)
    matrix_logp = MatrixNormal.dist(mu=mu, rcov=cov, lcov=lcov, shape=(q, p)).logp(x)
    f_matrix_logp = theano.function([cov, lcov, x], matrix_logp)
    matrix_logp_value = f_matrix_logp(cov_val, lcov_val, x_test)

    assert np.isclose(mv_logp_value, matrix_logp_value)
```
would that be something useful to add to the tests as well or would it be redundant? Also, I may have broken something with a `git pull` or something, since my PR is now cluttered with a bunch of other stuff. Is that normal?

File: pymc3/distributions/multivariate.py (hunk 7 above)
Comment (jordan-melendez) [reply pymc3/distributions/multivariate.py]: good points. If we allow for multiple observations, I do not know how to handle that in `random` and `_trquaddist`, where `matmul` and `solve_lower`, etc., will have to know that matrices are stacked somehow.

**Comment by aseyboldt** [comment]
@jordan-melendez Nope, that's not normal :-)
To fix it you can do something like that:
```
git checkout MatrixNormal
git fetch origin
git rebase -i origin/master
```
Then, in the editor delete commits you don't want to have in there. If you are very unlucky you might have to fix merge conflicts, but it tells you what you need to do if that happens.
If you push the result to github, you need to add `-f` or add a plus sign before the remote name to force it to overwrite the commits that are here currently.

**Comment by jordan-melendez** [comment]
@aseyboldt thank you! should be all fixed now.

As far as the argument labels go, should we match mine (`rcov`, `lcov`, etc.) with SciPy's? Their notation tends to follow the wikipedia article. Or does anyone have any better ideas? It's hard enough to keep this stuff straight without the notation making it worse. Even the descriptions of their actions ("among-row" in SciPy vs "within columns" in my code) is not particularly enlightening to me.

**Comment by junpenglao** [comment]
I think it is better to match with scipy

File: pymc3/distributions/multivariate.py (hunk 7 above)
Comment (bwengals) [reply pymc3/distributions/multivariate.py]: you're right.  on second thought, I could certainly be wrong, but the multiple observations thing is probably pretty rare.  Could punt on that one until someone opens an issue (maybe never?)  What do you think about raising a NotImplementedError if the length of `shape` is greater than 2 and saying something informative?

**Comment by bwengals** [comment]
looks like `matrix_normal` was new in version 0.17. @jordan-melendez thanks for fixing the PR.  RE the test, you could try adding it to the PR, then taking a look at the coverage report to help guide test writing.  Id favor trying to match the format/style + coverage of the existing tests for other dists, and doing something different if something specific about MatrixNormal turns out to make that a bad idea.  The testing looks a little complex so please feel free to ask about anything.

IMO the notation is OK either way, maybe prefer following scipy just for consistency.  I don't mind the rcov, lcov thing though, so long as the docstring makes things clear.

File: pymc3/distributions/multivariate.py (hunk 7 above)
Comment (jordan-melendez) [reply pymc3/distributions/multivariate.py]: I had hoped that the multiple observations case would be rare as well, so I'm okay leaving that alone for now. A NotImplementedError could be more helpful in the future so I'll try that in the next commit. Would it be feasible when using Theano to implement something like `_process_parameters` [here](https://github.com/scipy/scipy/blob/v0.18.1/scipy/stats/_multivariate.py#L785) that infers the qxp shape rather than requiring the shape argument? Then we could maybe allow for simplified entries for the various covariances, e.g., a 1d array could become the diagonal.

Comment (bwengals) [reply pymc3/distributions/multivariate.py]: Since there's one observation, I think that should work... but I'm not positive.  Maybe @aseyboldt could weigh in here?

Comment (aseyboldt) [reply pymc3/distributions/multivariate.py]: Not sure how exactly that would work. My guess it that you can't do this without quite significant changes in a couple of places. At the moment we assume that the shape of all free variables is statically known, and this might break assumption.

**Comment by jordan-melendez** [comment]
So I started adding MatrixNormal tests to test_distributions, but I do not have a good handle on what is going on in that file. I am not sure how to run my test code and debug it, so any help along those lines would be great.

**Comment by bwengals** [comment]
In order to merge, all the tests have to pass, and it's also important to have a good coverage score (maybe over 90%?).  A coverage report shows which lines of code actually get run by your tests.  If you have a coverage score of 50%, that means that your tests only caused 50% of the lines of code to actually be run.  Though it's an imperfect measure of how good your tests are, it's a useful tool to highlight issues you may have forgotten about.

In the summary below the comments here, take a look both at the Travis CI report and the coverage/coveralls (click "Details").  Take a close look at both reports (Travis report loads slowly..) and Travis will tell you which tests failed and why, and coveralls will show you which lines of code in the entire PyMC3 codebase are run by the tests.  Check the lines you've added for `MatrixNormal` in `distributions.py`.  Travis CI runs (all!) the tests in PyMC3 in several environments, python2 and python3, float32 and float64 precision, etc.  These tests start running everytime you `git push` to this PR.  Note that the coverage score might get a little messed up if the Travis tests fail.

To run tests locally, first install `pytest` and then run `pytest pymc3/tests/test_distributions.py`.  PyMC3 already contains fixtures for testing distributions, so what you did looks perfect to me, so long as tests are passing.  

Long story short, I think the tests you have look great.  I think it would be good to also modify [this test](https://github.com/pymc-devs/pymc3/blob/master/pymc3/tests/test_distributions_random.py#L533) for `MatrixNormal`.  Once all Travis tests pass and the coverage score is high I think it's good to go.

File: pymc3/distributions/multivariate.py [hunk 8]
```diff
@@ -1091,3 +1091,212 @@ def logp(self, x):
...
+        # Setup left covariance matrix
+        scale = pm.Lognormal('scale', mu=np.log(true_scale), sd=0.5)
+        rowcov = tt.nlinalg.diag([scale**(2*i) for i in range(m)])
+
+        vals = pm.MatrixNormal('vals', mu=mu, colchol=colchol, rowcov=rowcov,
+                               observed=data, shape=(m, n))
+    """
+
+    def __init__(self, mu=0, rowcov=None, rowchol=None, rowtau=None,
+                 colcov=None, colchol=None, coltau=None, *args, shape=None,
```
Comment (bwengals) [inline pymc3/distributions/multivariate.py]: The python 2 travis tests point to this line as causing an error:

    File "/home/travis/build/pymc-devs/pymc3/pymc3/distributions/multivariate.py", line 1183
        colcov=None, colchol=None, coltau=None, *args, shape=None,
                                                       ^
    SyntaxError: invalid syntax

I think it should be `shape=None, *args, **kwargs`.

**Comment by ColCarroll** [comment]
I love pytest flags, so let me add to @bwengals 's great advice: also install `pytest-cov` (or get all our dev requirements, `pip install -r requirements-dev.txt`), and use:
```
python -m pytest -xv --cov=pymc3/ --cov-report=html pymc3/tests/test_distributions.py
```

The `x` flag stops the tests after the first failure, `v` gives a little more information about what's happening, `-cov=pymc3` tells `coverage` what files to look at, and `--cov-report=html` tells `coverage` to generate a spiffy static website with your coverage!

After running that, you can `open htmlcov/index.html`, and navigate to the right file to see what has been covered.

File: pymc3/tests/test_distributions.py [hunk 9]
```diff
@@ -695,6 +704,29 @@ def test_mvnormal_init_fail(self):
             with pytest.raises(ValueError):
                 x = MvNormal('x', mu=np.zeros(3), cov=np.eye(3), tau=np.eye(3), shape=3)
 
+    @pytest.mark.parametrize('n', [1, 2, 3])
+    def test_matrixnormal(self, n):
+        self.pymc3_matches_scipy(MatrixNormal, RealMatrix(5, n),
```
Comment (bwengals) [inline pymc3/tests/test_distributions.py]: Travis shows:

    self = <pymc3.tests.test_distributions.TestMatchesScipy object at 0x7f786acb5e80>
    n = 1
        @pytest.mark.parametrize('n', [1, 2, 3])
        def test_matrixnormal(self, n):
    >       self.pymc3_matches_scipy(MatrixNormal, RealMatrix(5, n),
                                     {'mu': Vector(R, n), 'rowcov': PdMatrix(n),
                                      'colcov': PdMatrix(n)},
                                     matrix_normal_logpdf_cov)
    E       NameError: name 'MatrixNormal' is not defined

Don't forget `MatrixNormal` import at the top!

**Comment by jordan-melendez** [comment]
After the helpful advice, some tests now work! Unfortunately, float32 precision in (at least) Python 2.7 is close but not perfect. Any thoughts on what I could do to improve precision here?

**Comment by bwengals** [comment]
Since only the py2.7, float32 test fails, I think the results you compute are accurate.  Considering the magnitude of `logp` results,

    x: array(-3101513.3162162304)
    y: array(-3101491.012355871) 

the values are very similar.  Is it possible to tweak the mean/cov's you're using in the tests so that you get `logp` values with a smaller magnitude?

*Extremely* minor quibble:  There are some newlines throughout `MatrixNormal` that may not be necessary, like before `return`'s and in `__init__`.

After that I support merging.

**Comment by junpenglao** [comment]
Thanks @jordan-melendez! And congrats on the first contribution ;-)

**Comment by fonnesbeck** [comment]
Sorry to be late to the party. We need to get into the habit of adding any new functionality to the `RELEASE-NOTES.md` file as they are merged. I will quickly add a PR for it.

**Comment by bwengals** [comment]
Nicely done @jordan-melendez!
